### ISSUE TITLE
[BUG] - <title> Only one trial completes on Ax NAS

### ISSUE THREAD

**Issue by ekurtgl** [pytorch/tutorials#2495]
### Add Link

https://pytorch.org/tutorials/intermediate/ax_multiobjective_nas_tutorial.html

### Describe the bug

Hi,

I was able to get the tutorial notebook working, and now I am trying to implement Ax-based NAS on my own model. However, only one of the trials complete and all the others fail. I have one objective which is to maximize the val_accuracy. The training script runs fine without any problem when I run it on terminal as well. This is the error I am getting:

![image](https://github.com/pytorch/tutorials/assets/66868163/0bd80fc9-4fbf-4338-9181-e15692c3205f)

![image](https://github.com/pytorch/tutorials/assets/66868163/a320fc3f-4e86-4e85-9ad7-97735a4da732)

--------------
Full log:
---------------------------------------------------------------------------
FailureRateExceededError                  Traceback (most recent call last)
Cell In[10], line 1
----> 1 scheduler.run_all_trials()

File [~/anaconda3/envs/tpot/lib/python3.10/site-packages/ax/service/scheduler.py:999](https://file+.vscode-resource.vscode-cdn.net/home/emre/Desktop/NXP/NAS/tpot/src/~/anaconda3/envs/tpot/lib/python3.10/site-packages/ax/service/scheduler.py:999), in Scheduler.run_all_trials(self, timeout_hours, idle_callback)
    992 if self.options.total_trials is None:
    993     # NOTE: Capping on number of trials will likely be needed as fallback
    994     # for most stopping criteria, so we ensure `num_trials` is specified.
    995     raise ValueError(  # pragma: no cover
    996         "Please either specify `num_trials` in `SchedulerOptions` input "
    997         "to the `Scheduler` or use `run_n_trials` instead of `run_all_trials`."
    998     )
--> 999 for _ in self.run_trials_and_yield_results(
   1000     max_trials=not_none(self.options.total_trials),
   1001     timeout_hours=timeout_hours,
   1002     idle_callback=idle_callback,
   1003 ):
   1004     pass
   1005 return self.summarize_final_result()

File [~/anaconda3/envs/tpot/lib/python3.10/site-packages/ax/service/scheduler.py:899](https://file+.vscode-resource.vscode-cdn.net/home/emre/Desktop/NXP/NAS/tpot/src/~/anaconda3/envs/tpot/lib/python3.10/site-packages/ax/service/scheduler.py:899), in Scheduler.run_trials_and_yield_results(self, max_trials, ignore_global_stopping_strategy, timeout_hours, idle_callback)
    893         return
    895     yield self.wait_for_completed_trials_and_report_results(
    896         idle_callback, force_refit=True
    897     )
--> 899 yield self._complete_optimization(
    900     num_preexisting_trials=n_existing, idle_callback=idle_callback
    901 )
    902 return

File [~/anaconda3/envs/tpot/lib/python3.10/site-packages/ax/service/scheduler.py:1278](https://file+.vscode-resource.vscode-cdn.net/home/emre/Desktop/NXP/NAS/tpot/src/~/anaconda3/envs/tpot/lib/python3.10/site-packages/ax/service/scheduler.py:1278), in Scheduler._complete_optimization(self, num_preexisting_trials, idle_callback)
   1273 res = self.wait_for_completed_trials_and_report_results(
   1274     idle_callback=idle_callback, force_refit=True
   1275 )
   1276 # Raise an error if the failure rate exceeds tolerance at the
   1277 # end of the optimization.
-> 1278 self.error_if_failure_rate_exceeded(force_check=True)
   1279 self._record_run_trials_status(
   1280     num_preexisting_trials=num_preexisting_trials,
   1281     status=RunTrialsStatus.SUCCESS,
   1282 )
   1283 return res

File [~/anaconda3/envs/tpot/lib/python3.10/site-packages/ax/service/scheduler.py:779](https://file+.vscode-resource.vscode-cdn.net/home/emre/Desktop/NXP/NAS/tpot/src/~/anaconda3/envs/tpot/lib/python3.10/site-packages/ax/service/scheduler.py:779), in Scheduler.error_if_failure_rate_exceeded(self, force_check)
    771 if self._num_trials_bad_due_to_err > num_bad_in_scheduler [/](https://file+.vscode-resource.vscode-cdn.net/) 2:
    772     self.logger.warn(
    773         "MetricFetchE INFO: Sweep aborted due to an exceeded error rate, "
    774         "which was primarily caused by failure to fetch metrics. Please "
    775         "check if anything could cause your metrics to be flakey or "
    776         "broken."
    777     )
--> 779 raise self._get_failure_rate_exceeded_error(
    780     num_bad_in_scheduler=num_bad_in_scheduler,
    781     num_ran_in_scheduler=num_ran_in_scheduler,
    782 )

FailureRateExceededError: Failure rate exceeds the tolerated trial failure rate of 0.5 (at least 2 out of first 3 trials failed). Checks are triggered both at the end of a optimization and if at least 5 trials have failed.

-----------

![image](https://github.com/pytorch/tutorials/assets/66868163/b0103097-e93f-4bf0-8dac-12198a3884f3)

I don't set any objective thresholds. When I run the script from the terminal, it works fine every time, and val_accuracy never becomes NaN. What might be the reason for such behavior in trials? Thank you!

### Describe your environment

Ubuntu 20.04

**Comment by Balandat** [comment]
Hi @ekurtgl - hmm so the fundamental issue here seems to be that the trials are marked as failing, so the Scheduler stopping is the proper behavior. I'm not sure what makes the trials fail; and it's kind of hard to diagnose without more detail.

Looks like you're using a search space / model that is different from the one in the tutorial? It's kind of hard to say what's going wrong; have you tried the parameterizations that are marked as failed when run through the Scheduler manually? If those fail maybe it's parameterization dependent and you may want to narrow your search space around something you know works. If those don't fail then I wonder if it's some issue with the way things are setup that the logs are not properly separated and that causes the Scheduler issues when trying to read in the results.

**Comment by Balandat** [comment]
> I also have another question. Does Ax support trying different layers while performing the model search? I am thinking of something similar to [NNI's LayerChoice](https://nni.readthedocs.io/en/stable/reference/nas/search_space.html#layerchoice). Can ChoiceParameter be used for this purpose? If so, how should the parameter type be selected? Thank you in advance!

Yeah Ax is agnostic to what the choice parameter represents, so this should work. If it's a simple choice you may want to string encode this on the backend and then select the proper layer based on the string value of the choice parameter. But if the choice is actual numerical (e.g. like the sizes of the layers) it would make sense to encode those as numerical choice parameters. There is also some advanced functionality for hierarchical search spaces that can deal with nested choices (https://github.com/facebook/Ax/blob/main/ax/core/search_space.py#L399-L411) but we haven't really written docs/tutorials on that yet .

**Comment by ekurtgl** [comment]
> Hi @ekurtgl - hmm so the fundamental issue here seems to be that the trials are marked as failing, so the Scheduler stopping is the proper behavior. I'm not sure what makes the trials fail; and it's kind of hard to diagnose without more detail.
> 
> Looks like you're using a search space / model that is different from the one in the tutorial? It's kind of hard to say what's going wrong; have you tried the parameterizations that are marked as failed when run through the Scheduler manually? If those fail maybe it's parameterization dependent and you may want to narrow your search space around something you know works. If those don't fail then I wonder if it's some issue with the way things are setup that the logs are not properly separated and that causes the Scheduler issues when trying to read in the results.

Hi @Balandat ,

Yes, apparently some of the selections in the search space were causing inconsistent model architecture and the internal model code was throwing error. Thank you for your interest.
